File: Deliverables/HEAVYROUTE/src/main/java/com/heavyroute/core/dto/TripDTO.java
package com.heavyroute.core.dto;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Data Transfer Object (DTO) per la visualizzazione dei dettagli di un viaggio.
 * <p>
 * Questa classe rappresenta un "Read Model": è ottimizzata per la lettura da parte del client (Frontend/App).
 * A differenza dell'Entity, contiene dati "denormalizzati" o "appiattiti" (es. nomi autisti, modelli veicoli)
 * per evitare che il client debba effettuare chiamate aggiuntive ad altri microservizi o endpoint.
 * </p>
 */

@Data
@NoArgsConstructor
@AllArgsConstructor
public class TripDTO {

    /**
     * Identificativo tecnico del database.
     * Utile per operazioni di update/delete o per i link REST (HATEOAS).
     */
    private Long id;

    /**
     * Codice business leggibile.
     * È l'identificativo principale mostrato nelle interfacce utente.
     */
    private String tripCode;

    /**
     * Stato corrente del viaggio serializzato come stringa (es. "IN_PLANNING").
     * <p>
     * Viene esposto come String (e non come Enum) per disaccoppiare il contratto API:
     * se l'Enum interno cambia o viene rinominato, la stringa qui può essere mappata o mantenuta
     * per retrocompatibilità verso i client vecchi.
     * </p>
     */
    private String status;

    // --- DATI ARRICCHITI (ENRICHED DATA) ---

    /**
     * ID dell'autista (riferimento tecnico).
     * Corrisponde al campo 'driverId' dell'Entity.
     */
    private Long driverId;

    /**
     * Nome completo dell'autista (Dato arricchito).
     * <p>
     * <b>Nota:</b> Questo campo NON esiste nella tabella 'trips'.
     * Il Mapper deve recuperarlo interrogando il {@code DriverService} o la cache
     * usando il {@code driverId}.
     * </p>
     */
    private String driverName;

    /**
     * Targa del veicolo.
     */
    private String vehiclePlate;

    /**
     * Modello del veicolo (Dato arricchito).
     * <p>
     * Utile per la UI (es. "Iveco Stralis"), permette all'operatore di riconoscere
     * il mezzo senza dover decifrare la targa.
     * Da popolare tramite {@code VehicleService}.
     * </p>
     */
    private String vehicleModel;

    /**
     * Dettagli completi della Richiesta di Trasporto che ha originato questo viaggio.
     * <p>
     * <b>Struttura JSON:</b> Nel payload di risposta, questi dati saranno raggruppati
     * sotto la chiave {@code "request"}, permettendo al Frontend di passare questo oggetto
     * direttamente a componenti UI dedicati (es. {@code <RequestSummaryCard data={trip.request} />}).
     * </p>
     */
    private RequestDetailDTO request;
}

File: Deliverables/heavyroute_app/lib/features/planner/presentation/model/trip_dto.dart
import '../../../requests/models/request_detail_dto.dart';

/**
 * Data Transfer Object per la visualizzazione dei dettagli di un viaggio.
 * <p>
 * Mappa il "Read Model" del backend, includendo i dati arricchiti come
 * i nomi dei driver e i modelli dei veicoli per la visualizzazione in UI.
 * </p>
 * @author Roman
 */
class TripDTO {
  final int id;
  final String tripCode;
  final String status;
  final int? driverId;
  final String? driverName;
  final String? vehiclePlate;
  final String? vehicleModel;
  final RequestDetailDTO request;
  final int? clientId;
  final String? clientFullName;

  TripDTO({
    required this.id,
    required this.tripCode,
    required this.status,
    this.driverId,
    this.driverName,
    this.vehiclePlate,
    this.vehicleModel,
    required this.request,
    this.clientId,
    this.clientFullName,
  });

  /**
   * Crea un'istanza di {@link TripDTO} a partire da una mappa JSON.
   * @param json La mappa contenente i dati dal backend.
   * @return Un oggetto TripDTO formattato.
   */
  factory TripDTO.fromJson(Map<String, dynamic> json) {
    return TripDTO(
      id: json['id'],
      tripCode: json['tripCode'],
      status: json['status'],
      driverId: json['driverId'],
      driverName: json['driverName'],
      vehiclePlate: json['vehiclePlate'],
      vehicleModel: json['vehicleModel'],
      request: RequestDetailDTO.fromJson(json['request']),
      clientId: json['clientId'],
      clientFullName: json['clientFullName'],
    );
  }
}